src/main/java/in/org/projecteka/hiu/dataflow/HealthInfoManager.java

```diff
@@ -35,6 +45,62 @@
                                 consentDetail.get("hipName"))));
     }
 
+    public Flux<PatientDataEntry> fetchHealthInformation(List<String> dataRequestIds, String requesterId,
+                                                         int limit, int offset) {
+        return patientConsentRepository.fetchConsentRequestIds(dataRequestIds)
+                .collectList()
+                .flatMapMany(dataFlowRepository::fetchDataPartDetails)
+                .collectList()
+                .filter(dataParts -> isValidRequester(dataParts, requesterId))
+                .switchIfEmpty(Mono.error(unauthorizedRequester()))
+                .flatMapMany(dataParts -> getDataEntries(limit, offset, dataParts));
+    }
+
+    private boolean isValidRequester(List<DataPartDetail> dataParts, String requesterId) {
+        return dataParts.stream().allMatch(dataPart -> dataPart.getRequester().equals(requesterId));
+    }
+
+    private Flux<PatientDataEntry> getDataEntries(int limit, int offset, List<DataPartDetail> dataParts) {
+        HashMap<String, List<HealthInfoStatus>> dataPartStatuses = new HashMap<>();
+        HashMap<String, PatientDataEntry.PatientDataEntryBuilder> dataEntries = new HashMap<>();
```

Comment:
Do you really need this as a Map?

Reply:
Yeah, we need it as a map. If you can suggest something else that would be good.